TryFrom implementation on NameString

KSAlpha opened the pull request.

KSAlpha committed:
Update dependencies

KSAlpha committed:
Use standard TryFrom implementation

KSAlpha committed:
Satisfy clippy

KSAlpha committed:
Change lcow components download script

KSAlpha committed:
Add TLS version specifier

KSAlpha committed:
Use legacy docker (updated to 18.09)

KSAlpha committed:
Update dev-dependencies

Qwaz commented on src/problem.rs:
```diff
@@ -79,7 +79,7 @@ impl Manifest {
         executables
             .chain(readonly)
             .filter(|file_entry| file_entry.public())
```
FileEntry::public도 가능하지 않나요? filter와 map 하나의 스타일로 통일해주세요

KSAlpha replied:
사실 당연히(?) 해봤는데, 이해하기 어려운 에러가 나옵니다:
대신 `filter` 에 맞추게 되면, clippy가 불평하게 됩니다.

```
error[E0631]: type mismatch in function arguments
   --> src\problem.rs:81:14
    |
81  |             .filter(FileEntry::public)
    |              ^^^^^^ expected signature of `for<'r> fn(&'r &problem::configs::common::FileEntry) -> _`
    | 
   ::: src\problem\configs\common.rs:101:5
    |
101 |     pub fn public(&self) -> bool {
    |     ---------------------------- found signature of `for<'r> fn(&'r problem::configs::common::FileEntry) -> _`
```

Qwaz replied:
clippy 불평 메시지도 혹시 보여 주실 수 있나요

Qwaz replied:
filter는 인자가 `&Self::Item`이고 map은 인자가 `Self::Item`이라서 그렇네요!

KSAlpha replied:
아하... Clippy 불평은 다음과 같습니다.

```
error: redundant closure found
  --> src/problem.rs:82:18
   |
82 |             .map(|file_entry| file_entry.path())
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: remove closure as shown: `problem::configs::common::FileEntry::path`
   |
   = note: `-D clippy::redundant-closure` implied by `-D warnings`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure
```

Qwaz commented on src/lib.rs:
```diff
@@ -92,13 +93,25 @@ lazy_static! {
-    pub fn try_from(s: impl AsRef<str>) -> SomaResult<NameString> {
-        let s = s.as_ref();
-        if NAME_REGEX.is_match(s) {
+impl TryFrom<&str> for NameString {
```
`impl<T: Into<String>> TryFrom<T> for NameString`을 사용하는게 좋아 보입니다. 현재 두 개로 나뉘어 있는 구현을 하나로 합칠 수 있을뿐 아니라 `Cow<'a, str>` 등까지 추가로 커버됩니다.

Qwaz replied:
아래쪽에 반복돼 있는 PartialEq도 비슷하게 하나의 구현으로 어떻게 커버가 될 것 같은 기분이 들어요

KSAlpha replied:
대신 저렇게 하게되면 `str`처리가 상대적으로 불리해질 것이라고 생각합니다 (performance wise).
높은 확률로 `to_owned()`가 두번 불리게 될 것 같아요.

하지만 여기서 퍼포먼스 고려 안해도 될것 같으니 괜찮을 것 같습니다.

KSAlpha replied:
그런데 이 친구도 이해하기 어려운 에러를 주는군요:
```
error[E0119]: conflicting implementations of trait `std::convert::TryFrom<_>` for type `NameString`:
  --> src/lib.rs:96:1
   |
96 | impl<V: Into<String>> TryFrom<V> for NameString {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: conflicting implementation in crate `core`:
           - impl<T, U> std::convert::TryFrom<U> for T
             where U: std::convert::Into<T>;
```

Qwaz replied:
String / str 처리는 제가 찾아봤습니다. `Into<String>::into(String)`은 no-op이고 `Into<String>::into(&str)`은 `.to_owned()` 사용합니다.

Qwaz replied:
Trait Conflict는 예상치 못한 문제네요 ㅠㅠ 시도해 보시고 안 고쳐지면 지금 두 개로 나뉜 구현 유지하는 쪽으로 합시다

Qwaz replied:
이런 튜토리얼이 있는데 적용 가능한지 읽어보고 있습니다:
https://jwilm.io/blog/from-str-to-cow/

Qwaz commented on src/repository/manager.rs:
```diff
@@ -69,7 +70,7 @@ impl<'a> RepositoryManager<'a> {
         repo_name: impl AsRef<str>,
```
이 부분 고민을 좀 해봤는데, 저는 인자로 NameString을 받고 clap 쪽에서 파싱할 때 NameString을 들고 오는게 더 좋은 설계라고 느껴져요.

Qwaz replied:
https://docs.rs/clap/2.32.0/clap/macro.value_t.html

KSAlpha replied:
좋아보입니다.

Qwaz commented:
코드 좀 건드려 봤는데, clap 파싱 부분 말고는 개선할 수 있는게 없는 것 같네요 ㅠㅠ Rust의 Trait Conflict에 대한 공부가 되었습니다

KSAlpha committed:
Apply some CRs

KSAlpha committed:
Format code

KSAlpha committed:
Fix usage in tests

KSAlpha committed:
Move out NameString from location_to_backend

This fixes the error when the user add a repository of bad default name with compilant custom name.

Qwaz commented on src/bin/soma/main.rs:
```diff
@@ -17,7 +18,7 @@ mod terminal_printer;
 fn cli_env(data_dir: &mut DataDirectory) -> SomaResult<Environment<impl Connect, TerminalPrinter>> {
     Environment::new(
-        username().to_lowercase(),
+        NameString::try_from(username().to_lowercase())?,
```
그러고보니 username이 NameString이 아닌 사람은 Soma를 쓸 수 없군요 ㅠㅠ `to_lowercase` 말고 더 좋은 sanitize 방법을 나중에 생각해보는걸로

Qwaz commented on src/lib.rs:
```diff
@@ -110,21 +110,33 @@ impl TryFrom<String> for NameString {
+        &self.inner == other
+    }
+}
+
```
이건 왜 추가된건가요?

Qwaz replied:
지금 보니 이런 코드를 훔쳐와서(?) 써도 좋을 것 같네요.
https://doc.rust-lang.org/src/alloc/string.rs.html#1828-1853

KSAlpha replied:
테스트 중 `assert_eq!(NameString, &str)` 에서 요구됩니다. 저 매크로는 한번 시도해보겠습니다.

Qwaz commented on src/ops.rs:
```diff
@@ -16,26 +17,26 @@ use crate::problem::configs::SolidBinaryConfig;
     env: &mut Environment<impl Connect, impl Printer>,
     repo_location: &str,
-    repo_name: Option<&str>,
+    repo_name: Option<NameString>,
```
CLI 넣어주는 부분이 좀 덜 심플해지긴 하겠지만 add도 다른 오퍼레이션과 일관성 있게 `Option<&NameString>` 하는게 낫지 않을까요?

KSAlpha replied:
그럴까요? 대신 add 부분이 좀 많이 더러워질 수도 있습니다. 한번 바꿔볼테니 보고나서 결정하셔도 될듯 하네요.

Qwaz commented on src/repository/manager.rs:
```diff
@@ -155,7 +152,7 @@ impl<'a> RepositoryManager<'a> {
     }
 
     pub fn repo_exists(&self, repo_name: impl AsRef<str>) -> bool {
```
`repo_name: &NameString`나 `repo_name: TryInto<NameString>`?

KSAlpha replied:
`impl TryInto<NameString>`가 현재 구조에 가장 적합해보이네요, 시도해보겠습니다.

KSAlpha replied:
그럴려고 했으나, `&NameString`에 대한 처리가 애매해서, 그냥 &NameString으로 변경하였습니다.

Qwaz commented on tests/add_remove.rs:
```diff
@@ -13,18 +16,19 @@ fn test_add_remove() {
 
-    let repo_name = SIMPLE_BOF_REPO_NAME;
+    // Repository names should follow NameString rules
+    let repo_name = NameString::from_str(SIMPLE_BOF_REPO_NAME).unwrap();
```
common 모듈에 `NameString::from_str(...).unwrap()`으로 확장되는 함수나 매크로를 추가해도 좋을 것 같네요.

KSAlpha replied:
좋은 의견입니다.

KSAlpha commented:
변경 사항 확인 부탁드려요

KSAlpha committed:
Introduce GuaranteedSanitization in tests

KSAlpha committed:
`repo_exists` now requires `&NameString`

KSAlpha committed:
Use macro instead to implement PartialEq

KSAlpha committed:
Change ops::add function signature

Qwaz commented on src/bin/soma/commands/add.rs:
```diff
@@ -48,7 +48,8 @@ impl SomaCommand for AddCommand {
                 .map(NameString::from_str)
-                .transpose()?,
+                .transpose()?
+                .as_ref(),
```
와 이런게 있네

qwaz reviewed (approved):
LGTM
